### ISSUE TITLE
Cannot remove container stopped while in "Restarting"

### ISSUE THREAD

**Issue by Teudimundo** [docker/docker#22222]
**Output of `docker version`:**

```
Client:
 Version:      1.11.0
 API version:  1.23
 Go version:   go1.5.4
 Git commit:   4dc5990
 Built:        Wed Apr 13 18:34:23 2016
 OS/Arch:      linux/amd64

Server:
 Version:      1.11.0
 API version:  1.23
 Go version:   go1.5.4
 Git commit:   4dc5990
 Built:        Wed Apr 13 18:34:23 2016
 OS/Arch:      linux/amd64
```

**Output of `docker info`:**

```
Containers: 269
 Running: 13
 Paused: 0
 Stopped: 256
Images: 2994
Server Version: 1.11.0
Storage Driver: aufs
 Root Dir: /var/lib/docker/aufs
 Backing Filesystem: extfs
 Dirs: 3553
 Dirperm1 Supported: false
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: overlay bridge null host
Kernel Version: 3.13.0-85-generic
Operating System: Ubuntu 14.04.4 LTS
OSType: linux
Architecture: x86_64
CPUs: 8
Total Memory: 31.36 GiB
Name: xxxxxxxx
ID: xxx.xxx.xxx.xxx
Docker Root Dir: /var/lib/docker
Debug mode (client): false
Debug mode (server): false
Registry: https://index.docker.io/v1/
Cluster store: consul://xxx.xxx.xxx.xxx:8500
Cluster advertise: xxx.xxx.xxx.xxx:2376
```

**Additional environment details (AWS, VirtualBox, physical, etc.):**
Physical box

**Steps to reproduce the issue:**

Create a container with restart policy 'unless-stopped':

   `docker run -d --name restartbug --restart unless-stopped busybox /bin/true` 

checking the status I get:

   `748270c68181  busybox  "/bin/true"  25 seconds ago   Restarting (0) 10 seconds ago   restartbug`

After stopping it with `docker stop restartbug`:

    `748270c68181  busybox  "/bin/true" 57 seconds ago    Exited (0) 19 seconds ago   restartbug`

**Describe the results you received:**

Trying now to remove the container:

```
$ docker rm restartbug
Error response from daemon: Driver aufs failed to remove root filesystem 
748270c68181ca10d178efc89ad56bc6f54bf4129d652104f5d37faea0ad4a50: rename 
/var/lib/docker/aufs/diff/d20b5c2ac56e67f44a4b7771efc31d14d892bd2459a041561593f0af14d4ed5b 
/var/lib/docker/aufs/diff/d20b5c2ac56e67f44a4b7771efc31d14d892bd2459a041561593f0af14d4ed5b-removing: 
device or resource busy
```

and the status of the containers is then:

     `748270c68181  busybox  "/bin/true"   About a minute ago    Dead                 restartbug`

**Describe the results you expected:**
Being able to remove the container.

**Additional information you deem important (e.g. issue happens only occasionally):**

We have tried changing the `restart-policy` to `no` and it was working.

We also tried with `on-failure`:
1. `docker run -d --name restartbug2 --restart on-failure busybox /bin/false` 
2. `docker run -d --name restartbug3 --restart on-failure:1024 busybox /bin/false`
3. `docker run -d --name restartbug4 --restart on-failure:1 busybox /bin/false`

In the first 2 cases we issued the `stop` while in `Restarting` and we faced the same issue on the `rm`.
In the third case we waited for the container being in `Exit(1)` before removing the container, and in this
case it was successful.

**Comment by Teudimundo** [comment]
As additional information, and workaround if one realise the issue before doing the `rm`, executing:
`docker update --restart no restartbug5`

while in Restart allow then to `stop` (if required) and then remove the container without the problem.

**Comment by HackToday** [comment]
Confirmed hit that in master
```
Error response from daemon: Driver aufs failed to remove root filesystem 338b4ce0795f23da480056a8a6f010b41d48ef19c6111c35c59a803a062b5ca5: rename /var/lib/docker/aufs/diff/873b83c3d539130763caa5e92a6c360cd37ff755f015568a929048e4b1f1e5ca /var/lib/docker/aufs/diff/873b83c3d539130763caa5e92a6c360cd37ff755f015568a929048e4b1f1e5ca-removing: device or resource busy
```

but after `docker rm -f <container id>`, the container gone

**Comment by Teudimundo** [comment]
@HackToday, right, with `rm -f` you still get the error but the container is then removed. Thanks.

**Comment by mlaventure** [comment]
@Teudimundo do you have any running containerd mounting  `/var/run`? If so, it could be the same as #21969

**Comment by Teudimundo** [comment]
nope, no container is mounting `/var/run`.

**Comment by mlaventure** [comment]
@Teudimundo `/var/lib/docker`then?

I still can't find a way to reproduce this issue otherwise, I did come upon the issue that @tonistiigi #22237 PR is fixing but that's it.

I'll mark this as P3 until we have a dev that can reliably reproduce it for debugging.

**Comment by AleksandrIakhnev** [comment]
It easy to reproduce with Python 3.5 container : 
 docker run -d --name=test_name --restart=always python:3.5 python -c 'import badmodule'
docker stop test_name
docker rm test_name #fails with  error ' Driver aufs failed to remove root filesystem.... device or resource busy'
docker rm -f test_name #the same message, but container actually was removed

**Comment by thaJeztah** [comment]
@AleksandrIakhnev would you be able to test docker 1.11.1-rc1? I think this issue should be resolved, but if you are able to test it, that would be great! (note, it's a release candidate, so shouldn't be used in production) https://github.com/docker/docker/releases/tag/v1.11.1-rc1

**Comment by AleksandrIakhnev** [comment]
@thaJeztah Thanks for the anwer, I will try.

**Comment by AleksandrIakhnev** [comment]
@thaJeztah On  1.11.1-rc1  container can be removed without force flag and warnings, issue is fixed.

**Comment by thaJeztah** [comment]
@AleksandrIakhnev thanks so much! Good to hear it's resolved indeed; 1.11.1 GA should be released soon
